oasislmf/model_preparation/summaries.py

```diff
@@ -413,3 +419,155 @@ def generate_summaryxref_files(model_run_fp, analysis_settings):
+        if coverage_type in exposure_summary[peril]['all']['tiv_by_coverage']:
+            exposure_summary[peril]['all']['tiv_by_coverage'][coverage_type] += tiv_sum
+        else:
+            exposure_summary[peril]['all']['tiv_by_coverage'][coverage_type] = tiv_sum
+        exposure_summary[peril][status]['tiv'] += tiv_sum
+        exposure_summary[peril]['all']['tiv'] += tiv_sum
+
+    # Find number of locations
+    loc_count = df.loc[df['peril_id'] == peril_code, loc_num].drop_duplicates().count()
+    loc_count = int(loc_count)
+    exposure_summary[peril][status]['number_of_locations'] = loc_count
+    exposure_summary[peril]['all']['number_of_locations'] += loc_count
+
+    return exposure_summary
+
+
+@oasis_log
+def write_exposure_summary(
+    target_dir,
+    gul_inputs_df,
+    exposure_fp,
+    keys_errors_fp,
+    exposure_profile,
+    loc_num
+):
+    """
+    Create exposure summary as dictionary of TIVs and number of locations
+    grouped by peril and validity respectively. Writes dictionary as json file
+    to disk.
+
+    :param target_dir: directory on disk to write exposure summary file
+    :type target_dir: str
+
+    :param gul_inputs_df: dataframe from gul_inputs.get_gul_input_items(..)
+    :type gul_inputs_df: pandas.DataFrame
+
+    :param exposure_fp: file path to input exposure file
+    :type exposure_fp: str
+
+    :param keys_errors_fp: file path to keys erors file
+    :type keys_errors_fp: str
+
+    :param exposure_profile: profile defining exposure file
+    :type exposure_profile: dict
+
+    :param loc_num: location number column heading from exposure file
+    :type loc_num: str
+    """
+
+    # Get keys in GUL input items dataframe and merge with key errors to obtain
+    # status information
+    try:
+        gul_inputs_errors_df, _ = get_gul_input_items(
+            exposure_fp, keys_errors_fp, exposure_profile=exposure_profile
+        )
+    except OasisException:   # Empty dataframe
+        gul_inputs_errors_df = pd.DataFrame(columns=gul_inputs_df.columns)
+    keys_errors_df = pd.read_csv(keys_errors_fp)
+    keys_errors_df.columns = keys_errors_df.columns.str.lower()
+    gul_inputs_errors_df = merge_dataframes(
+        gul_inputs_errors_df,
+        keys_errors_df,
+        left_on=[loc_num, 'peril_id', 'coverage_type_id'],
+        right_on=['locid', 'perilid', 'coveragetypeid'],
+        how='left'
+    )
+    gul_inputs_errors_df.drop(
+        columns=['locid', 'perilid', 'coveragetypeid', 'message'],
+        axis=1,
+        inplace=True
+    )
+
+    # Compile summary of exposure data
+    exposure_summary = {}
+    for peril_code in gul_inputs_df['peril_id'].unique():
+        # Use descriptive names of perils as keys
+        for k, vals in PERILS.items():
+            if vals['id'] == peril_code:
+                peril = vals['desc']
+                break
+        exposure_summary[peril] = {}
```

Comment:
@hchagani-oasislmf This line 540

        exposure_summary[peril] = {}

breaks - the `peril` variable is defined inside the `for` loop over `PERILS`, but not outside. This causes the deterministic loss generation command to break (`oasislmf exposure run`).

Can you please test your branch version with FM3?

    oasislmf exposure run -s validation/examples/fm3

Reply:
Inserted `if not deterministic` condition before `write_exposure_summary` function call.